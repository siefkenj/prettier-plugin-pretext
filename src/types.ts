import type { IToken } from "chevrotain";
import type * as Prettier from "prettier";
import type {
    AttributeCstNode,
    ChardataCstNode,
    ContentCstNode,
    DocTypeDeclNode,
    DocumentCstNode,
    ElementCstNode,
    ExternalIDNode,
    PrologCstNode,
} from "@xml-tools/parser";

export type Fragment = {
    offset: number;
    printed: Doc;
    startLine?: number;
    endLine?: number;
    endOffset?: number;
    nodeType?: "element" | "text";
    tagName?: string | null;
};


// Override the children property of the content CST node because it is missing
// the PROCESSING_INSTRUCTION property.
interface ContentCstNodeExt extends Omit<ContentCstNode, "children"> {
    children: ContentCstNode["children"] & { PROCESSING_INSTRUCTION: IToken[] };
}

// Override the children property of the doctype CST node because it is missing
// the CLOSE property.
interface DocTypeDeclNodeExt extends Omit<DocTypeDeclNode, "children"> {
    children: DocTypeDeclNode["children"] & { CLOSE: IToken[] };
}

// Override the children property of the element CST node because its children
// object points to the content CST node that is missing the
// PROCESSING_INSTRUCTION property.
interface ElementCstNodeExt extends Omit<ElementCstNode, "children"> {
    children: Omit<ElementCstNode["children"], "content"> & {
        content: ContentCstNodeExt[];
    };
}

// Override the name property on the external ID CST node because it has the
// incorrect name value.
interface ExternalIDNodeExt extends Omit<ExternalIDNode, "name"> {
    name: "externalID";
}

// The type of elements that make up the given array T.
type ArrayElement<T> = T extends (infer E)[] ? E : never;

// A union of the properties of the given object that are arrays.
type ArrayProperties<T> = {
    [K in keyof T]: T[K] extends any[] ? K : never;
}[keyof T];

// A union of the properties of the given array T that can be used to index it.
// If the array is a tuple, then that's going to be the explicit indices of the
// array, otherwise it's going to just be number.
type IndexProperties<T extends { length: number }> = IsTuple<T> extends true
    ? Exclude<Partial<T>["length"], T["length"]>
    : number;

// Effectively performing T[P], except that it's telling TypeScript that it's
// safe to do this for tuples, arrays, or objects.
type IndexValue<T, P> = T extends any[]
    ? P extends number
        ? T[P]
        : never
    : P extends keyof T
    ? T[P]
    : never;

// Determines if an object T is an array like string[] (in which case this
// evaluates to false) or a tuple like [string] (in which case this evaluates to
// true).
// eslint-disable-next-line @typescript-eslint/no-unused-vars
type IsTuple<T> = T extends []
    ? true
    : T extends [infer First, ...infer Remain]
    ? IsTuple<Remain>
    : false;

type CallProperties<T> = T extends any[] ? IndexProperties<T> : keyof T;
type IterProperties<T> = T extends any[]
    ? IndexProperties<T>
    : ArrayProperties<T>;

type CallCallback<T, U> = (path: Path<T>, index: number, value: any) => U;
type EachCallback<T> = (
    path: Path<ArrayElement<T>>,
    index: number,
    value: any
) => void;
type MapCallback<T, U> = (
    path: Path<ArrayElement<T>>,
    index: number,
    value: any
) => U;

// This path interface is going to override a bunch of functions on the regular
// prettier AstPath interface. This is because we want stricter types than the
// current version of @types/prettier provides.
//
// For each of the tree walk functions (call, each, and map) this provides 5
// strict type signatures, along with a fallback at the end if you end up
// calling more than 5 properties deep (we don't do that but I've included it
// for completeness).
//
// getParentNode is being overridden because previously it was restricted to the
// type T of the given AST, but it's very unlikely you're going to be receiving
// a parent node that exactly matches your current node. So for now just
// returning any.
// prettier-ignore
interface StrictPath<T> {
  call<U>(callback: CallCallback<T, U>): U;
  call<U, P1 extends CallProperties<T>>(callback: CallCallback<IndexValue<T, P1>, U>, prop1: P1): U;
  call<U, P1 extends keyof T, P2 extends CallProperties<T[P1]>>(callback: CallCallback<IndexValue<IndexValue<T, P1>, P2>, U>, prop1: P1, prop2: P2): U;
  call<U, P1 extends keyof T, P2 extends CallProperties<T[P1]>, P3 extends CallProperties<IndexValue<T[P1], P2>>>(callback: CallCallback<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, U>, prop1: P1, prop2: P2, prop3: P3): U;
  call<U, P1 extends keyof T, P2 extends CallProperties<T[P1]>, P3 extends CallProperties<IndexValue<T[P1], P2>>, P4 extends CallProperties<IndexValue<IndexValue<T[P1], P2>, P3>>>(callback: CallCallback<IndexValue<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, P4>, U>, prop1: P1, prop2: P2, prop3: P3, prop4: P4): U;
  call<U, P extends PropertyKey>(callback: CallCallback<any, U>, prop1: P, prop2: P, prop3: P, prop4: P, ...props: P[]): U;

  each(callback: EachCallback<T>): void;
  each<P1 extends IterProperties<T>>(callback: EachCallback<IndexValue<T, P1>>, prop1: P1): void;
  each<P1 extends keyof T, P2 extends IterProperties<T[P1]>>(callback: EachCallback<IndexValue<IndexValue<T, P1>, P2>>, prop1: P1, prop2: P2): void;
  each<P1 extends keyof T, P2 extends IterProperties<T[P1]>, P3 extends IterProperties<IndexValue<T[P1], P2>>>(callback: EachCallback<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>>, prop1: P1, prop2: P2, prop3: P3): void;
  each<P1 extends keyof T, P2 extends IterProperties<T[P1]>, P3 extends IterProperties<IndexValue<T[P1], P2>>, P4 extends IterProperties<IndexValue<IndexValue<T[P1], P2>, P3>>>(callback: EachCallback<IndexValue<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, P4>>, prop1: P1, prop2: P2, prop3: P3, prop4: P4): void;
  each<P extends PropertyKey>(callback: EachCallback<any>, prop1: P, prop2: P, prop3: P, prop4: P, ...props: P[]): void;

  getParentNode: (count?: number | undefined) => any | null,

  map<U>(callback: MapCallback<T, U>): U[];
  map<U, P1 extends IterProperties<T>>(callback: MapCallback<IndexValue<T, P1>, U>, prop1: P1): U[];
  map<U, P1 extends keyof T, P2 extends IterProperties<T[P1]>>(callback: MapCallback<IndexValue<IndexValue<T, P1>, P2>, U>, prop1: P1, prop2: P2): U[];
  map<U, P1 extends keyof T, P2 extends IterProperties<T[P1]>, P3 extends IterProperties<IndexValue<T[P1], P2>>>(callback: MapCallback<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, U>, prop1: P1, prop2: P2, prop3: P3): U[];
  map<U, P1 extends keyof T, P2 extends IterProperties<T[P1]>, P3 extends IterProperties<IndexValue<T[P1], P2>>, P4 extends IterProperties<IndexValue<IndexValue<T[P1], P2>, P3>>>(callback: MapCallback<IndexValue<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, P4>, U>, prop1: P1, prop2: P2, prop3: P3, prop4: P4): U[];
  map<U, P extends PropertyKey>(callback: MapCallback<any, U>, prop1: P, prop2: P, prop3: P, prop4: P, ...props: P[]): U[];
}

// Exporting an overall node union type so that we can reference it in parsers
// and everything.
export type AnyNode =
    | AttributeCstNode
    | ChardataCstNode
    | ContentCstNodeExt
    | DocTypeDeclNodeExt
    | DocumentCstNode
    | ElementCstNodeExt
    | ExternalIDNodeExt
    | PrologCstNode;

// Reexporting every export from the parser so that the different node types can
// be referenced.
export * from "@xml-tools/parser";

// Reporting the IToken from chevrotain because that is effectively another
// node in th tree.
export type { IToken } from "chevrotain";

// Reexporting the Doc type mostly because it's annoying to have to reference
// type so deeply in the Prettier namespace. Also because we only really want to
// be pulling in types from this file as they're less likely to change.
export type Doc = Prettier.doc.builders.Doc;

// This is the same embed as is present in prettier, except that it's required.
export type Embed = (
    path: Path<AnyNode>,
    print: (path: Path<any>) => Doc,
    textToDoc: (text: string, options: Options) => Doc,
    options: Options
) => Doc | null;

// These are the regular options from prettier except they also include all of
// the options we defined in our plugin configuration.
export type Options = Prettier.ParserOptions<any> & {
    bracketSameLine: boolean;
    xmlSelfClosingSpace: boolean;
    xmlWhitespaceSensitivity: "ignore" | "strict";
};

// We're going to change the signature of our parse function to accept our
// options as opposed to the generic options that don't contain the options we
// defined in our plugin configuration.
export type Parser = Omit<Prettier.Parser<AnyNode>, "parse"> & {
    parse: (
        text: string,
        parsers: { [name: string]: Prettier.Parser<any> },
        options: Options
    ) => any;
};

// We're overwriting a bunch of function here that walk around the tree here
// because if you restrict the AST for the main path then presumably you're
// printing a lower node in the tree that won't match the current AST type.
export type Path<T> = Omit<Prettier.AstPath<T>, keyof StrictPath<T>> &
    StrictPath<T>;

// Overwriting the parsers option so that we can again pass our defined options
// into the parse functions.
export type Plugin = Omit<Prettier.Plugin<AnyNode>, "parsers" | "printers"> & {
    parsers: { [name: string]: Parser };
    printers: { [name: string]: Printer };
};

// We're overwriting the print function with a print function that accepts the
// options we've defined.
export type Printer = Omit<Prettier.Printer<AnyNode>, "embed" | "print"> & {
    print: (path: Path<AnyNode>, options: Options, print: Print) => Doc;
};

// This is the regular print node, except it's not restricted by the AST that
// is passed to the parent AST. That's because when you're using it, you are not
// typically printing the same node type again.
export type Print = (path: Path<any>) => Doc;