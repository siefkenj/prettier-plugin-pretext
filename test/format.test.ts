import { describe, expect, test } from "vitest";
import fs from "fs";
import path from "path";
import prettier from "prettier";

import { Options } from "../src/types";
import plugin from "../src/plugin";

const fixture = fs.readFileSync(path.join(__dirname, "./fixture.ptx"), "utf-8");
const example = fs.readFileSync(
    path.join(__dirname, "./example.ptx"),
    "utf-8"
);
const example2 = fs.readFileSync(
    path.join(__dirname, "./example2.ptx"),
    "utf-8"
);
const example3 = fs.readFileSync(
    path.join(__dirname, "./example3.ptx"),
    "utf-8"
);
const example4 = fs.readFileSync(
    path.join(__dirname, "./example4.ptx"),
    "utf-8"
);

async function format(content: string, opts: Partial<Options> = {}) {
    return prettier.format(content, {
        ...opts,
        parser: "ptx",
        plugins: [plugin as any as string], // hacky but it works
    });
}

test("defaults", async () => {
    const formatted = await format(fixture);
    expect(formatted).toMatchSnapshot();
});

test("basic pretext source", async () => {
    const formatted = await format(example);
    expect(formatted).toMatchSnapshot();
});


test("basic pretext source 2", async () => {
    const formatted = await format(example2);
    expect(formatted).toMatchSnapshot();
});


test("bracketSameLine => true", async () => {
    const formatted = await format(fixture, {
        bracketSameLine: true,
    });

    expect(formatted).toMatchSnapshot();
});

test("xmlSelfClosingSpace => false", async () => {
    const formatted =await format(fixture, {
        xmlSelfClosingSpace: false,
    });

    expect(formatted).toMatchSnapshot();
});

test("bracketSameLine => true, xmlSelfClosingSpace => false", async () => {
    const formatted = await format(fixture, {
        bracketSameLine: true,
        xmlSelfClosingSpace: false,
    });

    expect(formatted).toMatchSnapshot();
});

test("singleAttributePerLine => true", async () => {
    const formatted = await format(fixture, {
        singleAttributePerLine: true,
    });

    expect(formatted).toMatchSnapshot();
});

test("example3", async () => {
    const formatted = await format(example3);
    expect(formatted).toMatchSnapshot();
});

test("new", async  () => {
    const formatted = await format(example4);
    expect(formatted).toMatchSnapshot();
});

