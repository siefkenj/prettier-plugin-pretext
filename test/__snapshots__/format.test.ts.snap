// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic pretext source 1`] = `
"<section>
  <cline>abc(foo)</cline>
  <li xml:id="list-two-two-two-three">
    <title>Title on xref'ed list item</title>
     <p>Level 4, third.</p>
  </li>
  <subsection>
    <abstract>
      <p>
        This is a sample of many of the things you can do with <pretext />.
        Sometimes the math makes sense, sometimes it seems to be written in the
        first person, sort of like this Abstract.
      </p>
    </abstract>
  </subsection>
  <macros>
    \\newcommand{\\definiteintegral}[4]{\\int_{#1}^{#2}\\,#3\\,d#4} % this comment will be stripped
    \\newcommand{\\myequation}[2]{#1\\amp =#2} % testing alignment override
    % comments anywhere get stripped, with or without faux % comment \\% characters
    \\newcommand{\\indefiniteintegral}[2]{\\int#1\\,d#2}
    \\newcommand{\\testingescapedpercent}{ \\% } % just testing
  </macros>
  <macros>\\newcommand{\\definiteintegral}[4]{\\int_{#1}   }</macros>
</section>
"
`;

exports[`basic pretext source 2 1`] = `
"<section>
  <p>A video is a <br /> natural way to</p>
  <p>
    A video is a natural <br /> way to enhance a document when rendered in an
    electronic format, such as HTML web pages.
    It might be additional information that is hard to communicate with text
    (marine invertebrates swimming), a
  </p>
  <p>
    Some <trademark /> text <m>math and stuff</m>.
    and
    <ul>
      <li>foo</li>
      <li>bar</li>
    </ul>
     and more textsuch as HTML web pages.
    It might be additional information that is hard to communicate with text
    (marine invertebrates swimming), a
  </p>
  <sage>
    <output>fjdklsafjlkdjlkflk  fdjal f fjdlasjlf
     jfdajldlsldk fjlkfajslfjllkadkf                baz</output>
    <input>
                A = matrix(4,5, srange(20)) <br />a really really really long line i think that it is long<br />
                 A.rref()
    </input>
  </sage>
  <p>
    fun times
    <ul>
      <li>foo</li>
      <li>bar</li>
    </ul>
    with list
  </p>
  <p>
    fun times
    <ul>
      <li>foo</li>
      <li>bar</li>
    </ul>


    with list
  </p>
</section>
"
`;

exports[`bracketSameLine => true 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<?xml-model href="project.rnc" type="application/relax-ng-compact-syntax"?>
<!-- foo -->
<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  width="200"
  height="100"
  viewBox="0 0 200 100"
>
  <title>Style inheritance and the use element</title>

  <desc _attr="attr">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
  <?pagebreak?>

  <style />
  <style/>
  <style type="text/css">circle { stroke-opacity: 0.7; } .special circle {
  stroke: green; } use { stroke: purple; fill: orange; }</style>

  <yaml
    myveryveryveryverylongattributename="myveryveryveryverylongattributevalue"
  >- 1 - 2 - 3</style>

  <!-- inner comment -->

  <?pagebreak?>
  <g class="special" style="fill: blue">
    <circle id="c" cy="50" cx="50" r="40" stroke-width="20" />
  </g>
  <use xlink:href="#c" x="100" />
  <ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
  <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Sed at est eget enim consectetur accumsan.
  Aliquam pretium sodales ipsum quis dignissim.
  Sed id sem vel diam luctus fringilla.
  Aliquam quis egestas magna.
  Curabitur molestie lorem et odio porta, et molestie libero laoreet.
  Morbi rhoncus sagittis cursus.
  Nullam vehicula pretium consequat.
  Praesent porta ante at posuere sollicitudin.
  Nullam commodo tempor arcu, at condimentum neque elementum ut.</p>
  <span>content</span>

  <div>
    even more
    <content />
  </div>
</svg>
<!-- bar -->
"
`;

exports[`bracketSameLine => true, xmlSelfClosingSpace => false 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<?xml-model href="project.rnc" type="application/relax-ng-compact-syntax"?>
<!-- foo -->
<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  width="200"
  height="100"
  viewBox="0 0 200 100"
>
  <title>Style inheritance and the use element</title>

  <desc _attr="attr">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
  <?pagebreak?>

  <style />
  <style/>
  <style type="text/css">circle { stroke-opacity: 0.7; } .special circle {
  stroke: green; } use { stroke: purple; fill: orange; }</style>

  <yaml
    myveryveryveryverylongattributename="myveryveryveryverylongattributevalue"
  >- 1 - 2 - 3</style>

  <!-- inner comment -->

  <?pagebreak?>
  <g class="special" style="fill: blue">
    <circle id="c" cy="50" cx="50" r="40" stroke-width="20" />
  </g>
  <use xlink:href="#c" x="100" />
  <ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
  <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Sed at est eget enim consectetur accumsan.
  Aliquam pretium sodales ipsum quis dignissim.
  Sed id sem vel diam luctus fringilla.
  Aliquam quis egestas magna.
  Curabitur molestie lorem et odio porta, et molestie libero laoreet.
  Morbi rhoncus sagittis cursus.
  Nullam vehicula pretium consequat.
  Praesent porta ante at posuere sollicitudin.
  Nullam commodo tempor arcu, at condimentum neque elementum ut.</p>
  <span>content</span>

  <div>
    even more
    <content />
  </div>
</svg>
<!-- bar -->
"
`;

exports[`defaults 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<?xml-model href="project.rnc" type="application/relax-ng-compact-syntax"?>
<!-- foo -->
<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  width="200"
  height="100"
  viewBox="0 0 200 100"
>
  <title>Style inheritance and the use element</title>

  <desc _attr="attr">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
  <?pagebreak?>

  <style />
  <style/>
  <style type="text/css">circle { stroke-opacity: 0.7; } .special circle {
  stroke: green; } use { stroke: purple; fill: orange; }</style>

  <yaml
    myveryveryveryverylongattributename="myveryveryveryverylongattributevalue"
  >- 1 - 2 - 3</style>

  <!-- inner comment -->

  <?pagebreak?>
  <g class="special" style="fill: blue">
    <circle id="c" cy="50" cx="50" r="40" stroke-width="20" />
  </g>
  <use xlink:href="#c" x="100" />
  <ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
  <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Sed at est eget enim consectetur accumsan.
  Aliquam pretium sodales ipsum quis dignissim.
  Sed id sem vel diam luctus fringilla.
  Aliquam quis egestas magna.
  Curabitur molestie lorem et odio porta, et molestie libero laoreet.
  Morbi rhoncus sagittis cursus.
  Nullam vehicula pretium consequat.
  Praesent porta ante at posuere sollicitudin.
  Nullam commodo tempor arcu, at condimentum neque elementum ut.</p>
  <span>content</span>

  <div>
    even more
    <content />
  </div>
</svg>
<!-- bar -->
"
`;

exports[`example3 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<section xml:id="section1" label="mylabel">
  <title>My section</title>

  <idx><h>Index entry</h></idx>
  <idx><h>entries</h> <h>index</h></idx>
  <idx><h>my example</h><see>example, my</see></idx>

  <p>This <term>is</term> a shorter line.</p>

  <p>
    This is a paragraph with <alert>lots</alert> of <em>examples or such</em> of
    <term>inline</term> and <m>\\sin(x)</m> math-like <c>expressions</c>.
  </p>

  <p>
    We begin our study of algebraic structures by investigating sets associated
    with single operations that satisfy certain reasonable axioms; that is, we
    want to define an operation on a set in a way that will generalize such
    familiar structures as the integers <m>{\\mathbb Z}</m> together with the
    single operation of addition, or invertible <m>2 \\times 2</m> matrices
    together with the single operation of matrix multiplication.
    The integers and the <m>2 \\times 2</m> matrices, together with their
    respective single operations, are examples of algebraic structures known as
    groups.<fn>A test footnote, with no real content.</fn>
  </p>
  <p>
    We have already seen that two integers <m>a</m> and <m>b</m> are equivalent
    mod <m>n</m> if <m>n</m> divides <m>a - b</m>.
    The integers mod <m>n</m> also partition <m>{\\mathbb Z}</m> into <m>n</m>
    different equivalence classes; we will denote the set of these equivalence
    classes by
    <m>{\\mathbb Z}_n</m>. <notation>
      <usage>
        <m>\\mathbb Z_n</m>
      </usage>
      <description>the integers modulo <m>n</m></description>
    </notation> Consider the integers modulo 12 and the corresponding partition
    of the integers:<md>
      <mrow>{[0]} &amp; = \\{ \\ldots, -12, 0, 12, 24, \\ldots \\}</mrow>
      <mrow> {[1]} &amp; = \\{ \\ldots, -11, 1, 13, 25, \\ldots \\}</mrow>
      <mrow> &amp; \\vdots</mrow>
      <mrow> {[11]} &amp; = \\{ \\ldots, -1, 11, 23, 35, \\ldots \\}</mrow>
    </md>.
  </p>

  <example xml:id="example-groups-zn-arithmetic-laws">
    <title>Modular Arithmetic</title>

    <p>
      Most, but not all, of the usual laws of arithmetic hold for addition and
      multiplication in <m>{\\mathbb Z}_n</m>.
      For instance, it is not necessarily true that there is a multiplicative
      inverse.
      Consider the multiplication table for <m>{\\mathbb Z}_8</m> in
      <xref ref="figure-z8-mult" />.
      Notice that 2, 4, and 6 do not have multiplicative inverses; that is, for <m>n
      = 2</m>, 4, or 6, there is no integer <m>k</m> such that
      <m>k n \\equiv 1 \\pmod{ 8}</m>.
    </p>

    <figure xml:id="figure-z8-mult">
      <caption>Multiplication table for <m>{\\mathbb Z_8}</m></caption>
      <p>
        <me>\\begin{array}{c|cccccccc}
                    <![CDATA[\\cdot & 0 & 1 & 2 & 3 & 4 & 5 & 6 & 7 \\\\
                    \\hline
                    0       & 0 & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\\\
                    1       & 0 & 1 & 2 & 3 & 4 & 5 & 6 & 7 \\\\
                    2       & 0 & 2 & 4 & 6 & 0 & 2 & 4 & 6 \\\\
                    3       & 0 & 3 & 6 & 1 & 4 & 7 & 2 & 5 \\\\
                    4       & 0 & 4 & 0 & 4 & 0 & 4 & 0 & 4 \\\\
                    5       & 0 & 5 & 2 & 7 & 4 & 1 & 6 & 3 \\\\
                    6       & 0 & 6 & 4 & 2 & 0 & 6 & 4 & 2 \\\\
                    7       & 0 & 7 & 6 & 5 & 4 & 3 & 2 & 1 ]]>
                    \\end{array}</me>
      </p>
    </figure>
  </example>

  <p>
    This is a paragraph with <alert>lots and lots</alert> of <em>examples</em> of
    <term>inline</term> and <m>\\sin(x)</m> math-like <c>expressions</c>.
  </p>

  <p>
    Some display math: <me>\\int_1^2 f(x) \\, dx</me>.
    How does that look?
  </p>

  <proposition xml:id="proposition-zn-equiv-classes">
    <statement>
      <p>
        Let <m>{\\mathbb Z}_n</m> be the set of equivalence classes of the
        integers mod <m>n</m> and <m>a, b, c \\in {\\mathbb Z}_n</m>.
        <ol>
          <li>
            <p>
              Addition and multiplication are commutative:<md>
                <mrow> a + b  &amp; \\equiv  b + a \\pmod{ n}</mrow>
                <mrow> a  b   &amp; \\equiv  b  a \\pmod{ n}. </mrow>
              </md>
            </p>
          </li>

          <li>
            <p>
              Addition and multiplication are associative:<md>
                <mrow> (a + b) + c  &amp; \\equiv  a + (b + c)\\pmod{ n}</mrow>
                <mrow> (a  b)  c    &amp; \\equiv  a   (b  c) \\pmod{ n}</mrow>
              </md>.
            </p>
          </li>

          <li>
            <p>
              There are both additive and multiplicative identities:<md>
                <mrow> a + 0  &amp; \\equiv  a \\pmod{ n}</mrow>
                <mrow> a \\cdot  1  &amp; \\equiv  a \\pmod{ n}</mrow>
              </md>.
            </p>
          </li>

          <li>
            <p>
              Multiplication distributes over addition:<md>
                <mrow>a (b + c) \\equiv a b + a c \\pmod{ n}</mrow>
              </md>.
            </p>
          </li>

          <li>
            <p>
              For every integer <m>a</m> there is an additive inverse <m>-a</m>:<md>
                <mrow>a + (-a) \\equiv 0 \\pmod{ n}</mrow>
              </md>.
            </p>
          </li>

          <li>
            <p>
              Let <m>a</m> be a nonzero integer.
              Then <m>\\gcd(a,n) = 1</m> if and only if there exists a
              multiplicative inverse <m>b</m> for <m>a \\pmod{n}</m>; that is, a
              nonzero integer
              <m>b</m> such that <me>a b \\equiv 1 \\pmod{ n}</me>.
            </p>
          </li>
        </ol>

      </p>
    </statement>

    <proof>
      <p>We will prove (1) and (6) and leave the remaining properties to be
      proven in the exercises.</p>

      <p>
        (1) Addition and multiplication <latex /> are commutative modulo <m>n</m>
        since the remainder of <m>a + b</m> divided by <m>n</m> is the same as
        the remainder of <m>b + a</m> divided by <m>n</m>.
      </p>

      <p>
        (6) Suppose that <m>\\gcd(a, n) = 1</m>.
        Then there exist integers <m>r</m> and <m>s</m> such that
        <m>ar + ns = 1</m>.
        Since <m>ns = 1 - ar</m>, it must be the case that
        <m>ar \\equiv 1 \\pmod{n}</m>.
        Letting <m>b</m> be the equivalence class of <m>r</m>,
        <m>a b \\equiv 1\\pmod{n}</m>.
      </p>

      <p>
        Conversely, suppose that there exists an integer <m>b</m> such that <m>ab
        \\equiv 1 \\pmod{ n}</m>.
        Then <m>n</m> divides <m>ab -1</m>, so there is an integer <m>k</m> such
        that <m>ab - nk = 1</m>.
        Let <m>d = \\gcd(a,n)</m>.
        Since <m>d</m> divides <m>ab - nk</m>, <m>d</m>
        must also divide 1; hence, <m>d = 1</m>.
      </p>
    </proof>
  </proposition>
</section>
"
`;

exports[`new 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<section xml:id="section1" label="mylabel">
  <title>My section about <pretext /></title>


  <p>
    short <me>\\sin(x) + 2</me>, paragraph?
    Another sentence.
  </p>
</section>
"
`;

exports[`singleAttributePerLine => true 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<?xml-model href="project.rnc" type="application/relax-ng-compact-syntax"?>
<!-- foo -->
<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  width="200"
  height="100"
  viewBox="0 0 200 100"
>
  <title>Style inheritance and the use element</title>

  <desc _attr="attr">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
  <?pagebreak?>

  <style />
  <style/>
  <style type="text/css">circle { stroke-opacity: 0.7; } .special circle {
  stroke: green; } use { stroke: purple; fill: orange; }</style>

  <yaml
    myveryveryveryverylongattributename="myveryveryveryverylongattributevalue"
  >- 1 - 2 - 3</style>

  <!-- inner comment -->

  <?pagebreak?>
  <g
    class="special"
    style="fill: blue"
  >
    <circle
      id="c"
      cy="50"
      cx="50"
      r="40"
      stroke-width="20" />
  </g>
  <use
    xlink:href="#c"
    x="100" />
  <ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
  <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Sed at est eget enim consectetur accumsan.
  Aliquam pretium sodales ipsum quis dignissim.
  Sed id sem vel diam luctus fringilla.
  Aliquam quis egestas magna.
  Curabitur molestie lorem et odio porta, et molestie libero laoreet.
  Morbi rhoncus sagittis cursus.
  Nullam vehicula pretium consequat.
  Praesent porta ante at posuere sollicitudin.
  Nullam commodo tempor arcu, at condimentum neque elementum ut.</p>
  <span>content</span>

  <div>
    even more
    <content />
  </div>
</svg>
<!-- bar -->
"
`;

exports[`xmlSelfClosingSpace => false 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>

<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<?xml-model href="project.rnc" type="application/relax-ng-compact-syntax"?>
<!-- foo -->
<svg
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
  width="200"
  height="100"
  viewBox="0 0 200 100"
>
  <title>Style inheritance and the use element</title>

  <desc _attr="attr">
    &anp; &#12345;
    <![CDATA[
      foo
    ]]>
    bar
  </desc>
  <?pagebreak?>

  <style />
  <style/>
  <style type="text/css">circle { stroke-opacity: 0.7; } .special circle {
  stroke: green; } use { stroke: purple; fill: orange; }</style>

  <yaml
    myveryveryveryverylongattributename="myveryveryveryverylongattributevalue"
  >- 1 - 2 - 3</style>

  <!-- inner comment -->

  <?pagebreak?>
  <g class="special" style="fill: blue">
    <circle id="c" cy="50" cx="50" r="40" stroke-width="20" />
  </g>
  <use xlink:href="#c" x="100" />
  <ignored>
    <!-- prettier-ignore-start -->
      <   ignored   />
    <!-- prettier-ignore-end -->
  </ignored>
  <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Sed at est eget enim consectetur accumsan.
  Aliquam pretium sodales ipsum quis dignissim.
  Sed id sem vel diam luctus fringilla.
  Aliquam quis egestas magna.
  Curabitur molestie lorem et odio porta, et molestie libero laoreet.
  Morbi rhoncus sagittis cursus.
  Nullam vehicula pretium consequat.
  Praesent porta ante at posuere sollicitudin.
  Nullam commodo tempor arcu, at condimentum neque elementum ut.</p>
  <span>content</span>

  <div>
    even more
    <content />
  </div>
</svg>
<!-- bar -->
"
`;
